sdks/python/apache_beam/coders/union_coder.py

```diff
@@ -0,0 +1,94 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pytype: skip-file
+
+import struct
+
+from typing import Iterable, List
+
+from apache_beam.coders import Coder
+from apache_beam.coders.coders import FastCoder
+from apache_beam.typehints import typehints
+
+
+class UnionCoder(FastCoder):
+  def __init__(self, components):
+    # type: (Iterable[Coder]) -> None
+    if not components or not isinstance(components, list):
+      raise ValueError('A valid list of Coders must be provided.')
+
+    if len(components) > 255 or len(components) <= 1:
+      raise ValueError(
+          'The number of components for UnionCoder must be between 2 and 255.')
+
+    self._coders = components
+    self._coder_typehints = {
+        c.to_type_hint(): (struct.pack("B", i), c)
+        for i,
+        c in enumerate(self._coders)
+    }
+
+  def encode(self, value) -> bytes:
+    """
+        Encodes the given Union value into bytes.
+        """
+    typehint_type = type(value)
+    coder_t = self._coder_typehints.get(typehint_type, None)
+    if coder_t:
+      return coder_t[0] + coder_t[1].encode(value)
+    else:
+      raise ValueError(
+          "Unknown type {} for UnionCoder with the value {}. ".format(
+              typehint_type, value))
+
+  def decode(self, encoded: bytes):
+    """
+        Decodes the given bytes into a Union value.
+        """
+    try:
+      coder_index = struct.unpack("B", encoded[:1])[0]
+      coder = self._coders[coder_index]
+
+      return coder.decode(encoded[1:])
+    except Exception:  # pylint: disable=broad-except
+      raise ValueError(f"cannot decode {encoded}")
+
+  def is_deterministic(self) -> bool:
+    """
+        Returns True if all sub-coders are deterministic.
+        """
+    return all(c.is_deterministic() for c in self._coders)
+
+  def to_type_hint(self) -> typehints.UnionConstraint:
+    """
+        Returns a type hint representing the Union type with the sub-coders.
+        """
+    return typehints.Union[list(self._coder_typehints.keys())]
+
+  def coders(self):
+    # type: () -> List[Coder]
+    return self._coders
+
+  def __eq__(self, other):
+    return type(self) == type(other) and self._coders == other.coders()
+
+  def __repr__(self) -> str:
```

Comment:
do we need hash? TupleCoder has more APIs. Do we need to cover more?

Reply:
The URN is needed if you intend this coder to be understandable outside this SDK (e.g. by runners or other SDKs). In this case, however, we'd probably want a formal spec and consider whether the coders should be ordered.

Reply:
@robertwb and I talked about this. For now, we do not need URN since we are not sure whether this UnionCoder is largely useful across SDKs.

Reply:
Do we need to create an impl for this coder since it's a FastCoder?

Reply:
I think this boils down to whether we want to Cythonize this coder (to make it faster and implement in C). We either do that and create an impl + move implementation of encode/decode to an impl, or we inherit this class from Coder and not FastCoder.

For more info, see:
https://github.com/apache/beam/blob/9fdc59b2c0f9a54457d9cde3117e9f0c104d59f1/sdks/python/apache_beam/coders/coders.py

Reply:
Thanks for pointing this out. I missed that diff. For now, I think it is better to inherit `Coder` for now.